### ISSUE TITLE
Replace C-extension with pure Ruby

### ISSUE THREAD

**Issue by nateberkopec** [puma/puma#1889]
@ioquatix is getting good results with async-http https://github.com/socketry/async-http. 

It's 2019 and the JIT is Becoming a Thing in CRuby, which makes me think that the marginal gains of using the old Mongrel parser is probably outweighed by the simplicity of removing our native extensions.

This would also greatly improve the Windows experience.

**Comment by nateberkopec** [comment]
FWIW I mean the HTTP parsing part. We may be stuck on the SSL part, but that's a separate issue.

**Comment by evanphx** [comment]
No, we should not do this. The JIT'd code will never be as fast as the current code and we get nothing expect new bugs from doing it.

**Comment by nateberkopec** [comment]
I don't think it will be faster, I just think it would be Fast Enough. Adding 1ms of latency to every request would be worth getting rid of C IMO.

But the fact that this is decade(?)-old battle-tested code is of course true and not to be treated lightly.

**Comment by ioquatix** [comment]
Writing code in C has its own limitations but maybe I should clarify my position.

Parsing HTTP/1 is not that hard. But doing it robustly can be. I’ve done both - a Ragel parser in C and pure Ruby implementation which is done by hand and is fast enough but probably not 100% correct. My plan is to eventually either deprecate HTTP/1 or use Ragel... but the current implementation is fine for now.

I would say if you have working code don’t bother changing it... but it’s definitely a limitation if you want to support the JIT, JRuby, TruffleRuby.

As pointed out, C is definitely faster and less memory intensive, but parsing the HTTP/1 request is probably less than 1ms, and probably way less... raw response bodies with content length or connection close are almost zero overhead, chunked encoding is a bit more overhead but most of that ends up invoking native code anyway...

**Comment by evanphx** [comment]
If someone can show a pure-ruby http parser that only adds 1ms to requests I'll certainly entertain the idea.

**Comment by nateberkopec** [comment]
.@ioquatix's stuff could be used to a "spike" a test: https://github.com/socketry/protocol-http and https://github.com/socketry/protocol-http1.

**Comment by boazsegev** [comment]
Late to the party, but I would add that C improves memory usage over time, not just speed (AFAIK).

This is related to the fact that the C parser doesn't produce as many side effects (temporary objects) as a Ruby implementation is likely to require.

When writing a parser using Ruby code, it's very easy to end up with a bunch of temporary objects / strings that later require the GC to clean them up. I believe this had been optimized a lot with Ruby... but I doubt it could match C in this respect.

**Comment by dentarg** [comment]
> This would also greatly improve the Windows experience

And the JRuby experience, as the implementations can go out of sync (see https://github.com/puma/puma/issues/1978 for an example)

**Comment by ianks** [comment]
would it make sense to survey the lay of the land of modern http parsers? the performance benefits would probably be negligible, but there are potential benefits w/r/t security and maintainability.

- https://github.com/h2o/picohttpparser
- https://github.com/nodejs/http-parser

**Comment by eregon** [comment]
FWIW, I think at some point @aardvark179 experimented with a Ruby version of the Puma parser, by generating Ruby code from Ragel. That was actually quite a bit faster than the C extension on TruffleRuby at the time IIRC, probably in part because Strings in C extensions were not very fast at the time.
Another part that Ruby Strings use a Rope representation on TruffleRuby, which depending on the operations can be quite a bit faster.
Also parsing in C means migrating from a malloc()-ed `char*` to Ruby Strings, which means one copy. The copy might be possible to avoid by using the `char*` of a `RSTRING_PTR()` to parse, but then things like computing the code range is done lazily and might be more expensive.

HTTP headers seems fairly easy to parse, so I think it would be interesting to experiment with this.

Also, Ruby code can be executed in parallel with JRuby/TruffleRuby/Ractor, but C extension code currently not.
It's becoming possible to mark C extensions as Ractor-safe or thread-safe though: https://bugs.ruby-lang.org/issues/17307

**Comment by ioquatix** [comment]
You can use the following gems to implement a pure Ruby HTTP parser:

- https://github.com/socketry/protocol-http1
- https://github.com/socketry/protocol-http2

**Comment by derekargueta** [comment]
Chipping in late here, I think another good candidate to consider would be [llhttp](https://github.com/nodejs/llhttp). Node's C http-parser project is deprecated and won't be supported any more, llhttp is the replacement for it. It's getting good mileage in the Node runtime, but will also become the HTTP/1.x parser [for the Envoy proxy project](https://github.com/envoyproxy/envoy/pull/15814). Ruby MRI bindings have been created for it: https://github.com/metabahn/llhttp

This has a few benefits
* a shared implementation that this project alone doesn't have to maintain
* use in other projects (Node, Envoy) give it a lot of miles
* although it's C code, that C code is generated from a more maintainable Typescript implementation

**Comment by dentarg** [comment]
What about using [Rust](https://github.com/hyperium/hyper)? Would it be possible? I've heard [it is](https://github.com/rubygems/rubygems/pull/6149) a [thing](https://github.com/matsadler/magnus) now. :)

**Comment by evanphx** [comment]
What advantage would there be to rewriting it in Rust?

**Comment by boazsegev** [comment]
If I may weigh in, again:

I can see very little up-side to having only the parser (and not the reactor) written in a lower-level language whereas I can see a big maintenance up-side for having a Ruby HTTP parser.

The small advantage of having the parser written as an extension is the ability to avoid temporary objects and memory fragmentation. The Ruby garbage collector takes more time to clear out the temporary objects, which in turn increases memory fragmentation and impacts performance. However, the Ruby runtime steadily improved in this regard and with the current JIT it is probably efficient enough.

On the negative side, unless we release the Ruby MRI global instruction lock (GIL) and move both the reactor and the parser to an truly parallel IO thread, than there will be no real performance gain and quite a lot of additional work in having / keeping a non-Ruby parser.

A new (non-Ruby) parser will run the risk of introducing new security issues and bugs. Also, future maintenance will require Rust / C programmers which are interested in maintaining this Ruby project (a challenge enough for the existing Java / C parts).

IMHO, the added maintenance cost isn't worth the possible gains (unless going all in) whereas a completely Ruby code base would be an added bonus for maintenance.

**Comment by ianks** [comment]
> A new (non-Ruby) parser will run the risk of introducing new security issues and bugs.

I'm not so sure about this. Sure, writing a _brand new_ native parser would likely cause huge security issues, but porting in a battle-tested one (such as [`httparse`](https://github.com/seanmonstar/httparse) would likely be much more safe.

**Comment by dentarg** [comment]
> The JIT'd code will never be as fast as the current code

Time to revisit this statement? https://railsatscale.com/2023-08-29-ruby-outperforms-c/

**Comment by MSP-Greg** [comment]
> Time to revisit this statement?

I've been thinking about this, but it would be helpful to push the minimum Ruby version to at least 2.7.  Puma has added a lot more tests for request validation and Ruby is more robust.

Considering I have links to old and new RFC's on my desktops and tablet, and I've worked a bit with SSL, I think this might work.

It would certainly with help with issues in the future, as changes to the current code may require changes of c, java, and possibly ragel code. And of course, the future with http/2 and/or http/3...

**Comment by evanphx** [comment]
I'd say the only reason to revisit it is if there are real, tangible benefits (better performance, better platform support, something). Put another way, the only way we'd revisit this is if someone does the work and then we can evaluate it's working performance. If someone wants to go down that path, I'm all for considering it at that point.

**Comment by nateberkopec** [comment]
To any hero who wishes to do this: I think certainly at least a first pass can just re-use [protocol](https://github.com/socketry/protocol-http1) from @ioquatix (or some other pure-ruby implementation that already exists, I don't really care). 

As Evan said, a working concept that has some real tangible benefit is more important than the details of how it's done or what dependency is added, which we can sort out later.

**Comment by swebb** [comment]
As an exercise I took a stab at writing a new parser in ruby. Just as a fun/learning exercise, not in any real expectation it will get merged into Puma. I took inspiration from [Tenderlove's blog post](https://tenderlovemaking.com/2023/09/02/fast-tokenizers-with-stringscanner/) and used StringScanner. The parser is only about 130 LOC and I haven't optimised it at all yet, but it does manage to pass most of the Puma test suite. I'm not quite ready to share it publicly yet, but I'll post again when it is.

**Comment by swebb** [comment]
I took a second attempt at this in #3815 using [protocol-http1](https://github.com/socketry/protocol-http1).
